src/tabs/Pixnflix/index.tsx

```diff
@@ -1,57 +1,252 @@
+import { Button, ButtonGroup, Divider, NumericInput } from '@blueprintjs/core';
+import { IconNames } from '@blueprintjs/icons';
+import {
+  ErrorLogger,
+  DEFAULT_WIDTH,
+  DEFAULT_HEIGHT,
+  DEFAULT_FPS,
+} from '../../bundles/pix_n_flix/types';
 
 type Props = {
   children?: never;
   className?: string;
   debuggerContext: any;
 };
 
-type State = {};
+type SideContentVideoDisplayMode = 'video' | 'still';
+
+type State = {
+  width: number;
+  height: number;
+  FPS: number;
+  mode: SideContentVideoDisplayMode;
+};
+
+type PixNFlix = {
+  toReplString: () => string;
+  init: (
+    video: HTMLVideoElement | null,
+    canvas: HTMLCanvasElement | null,
+    errorLogger: ErrorLogger
+  ) => void;
+  deinit: () => void;
+  startVideo: () => void;
+  snapPicture: () => void;
+  updateFPS: (fps: number) => void;
+  updateDimensions: (width: number, height: number) => void;
+};
 
 class Repeat extends React.Component<Props, State> {
   private $video: HTMLVideoElement | null = null;
 
   private $canvas: HTMLCanvasElement | null = null;
 
+  private pixNFlix: PixNFlix;
+
   constructor(props: any) {
     super(props);
-    this.state = {};
+    this.state = {
+      width: DEFAULT_WIDTH,
+      height: DEFAULT_HEIGHT,
+      FPS: DEFAULT_FPS,
+      mode: 'video' as SideContentVideoDisplayMode,
+    };
+    const { debuggerContext } = this.props;
+    this.pixNFlix = debuggerContext.result.value;
   }
 
   public componentDidMount() {
+    this.setupVideoService();
+    window.addEventListener('beforeunload', this.pixNFlix.deinit);
+  }
+
+  public componentWillUnmount() {
+    this.closeVideo();
+    window.removeEventListener('beforeunload', this.pixNFlix.deinit);
+  }
+
+  public setupVideoService = () => {
     if (this.$video && this.$canvas) {
-      // eslint-disable-next-line react/destructuring-assignment
-      this.props.debuggerContext.result.value.init(
-        this.$video,
-        this.$canvas,
-        this.printError
-      );
+      const { debuggerContext } = this.props;
+      this.pixNFlix = debuggerContext.result.value;
+      this.pixNFlix.init(this.$video, this.$canvas, this.printError);
     }
-  }
+  };
+
+  public closeVideo = () => {
+    this.pixNFlix.deinit();
+  };
 
-  public printError = () => {};
+  public handleStartVideo = () => {
+    this.pixNFlix.startVideo();
+  };
+
+  public handleSnapPicture = () => {
+    this.pixNFlix.snapPicture();
+  };
+
+  public handleWidthChange = (n: number) => {
+    if (n > 0 && n <= 500) {
+      this.setState((prevState) => ({
+        width: n,
+        height: prevState.height,
+      }));
+      const { height } = this.state;
+      this.handleUpdateDimensions(n, height);
+    }
+  };
+
+  public handleHeightChange = (m: number) => {
+    if (m > 0 && m <= 500) {
+      this.setState((prevState) => ({
+        width: prevState.width,
+        height: m,
+      }));
+      const { width } = this.state;
+      this.handleUpdateDimensions(width, m);
+    }
+  };

```

Comment:
Can `handleWidthChange` and `handleHeightChange` be combined into one function instead? I find that this doesnt really obey the DRY (Do not Repeat Yourself) principle...

Reply:
Yeah, I agree it's a little repetitive. This was done because NumericInput's onValueChange requires a function that only takes in a single argument, so unfortunately we cannot just use a single handleUpdateDimension function that takes in both the height and width.